src/OpenTelemetry/CompositeProcessor.cs

```diff
@@ -24,7 +24,7 @@ namespace OpenTelemetry
 {
     public class CompositeProcessor<T> : BaseProcessor<T>
     {
-        private DoublyLinkedListNode head;
+        private readonly DoublyLinkedListNode head;
```

Comment:
I want to get more perspectives regarding this one.
Making it `readonly` makes me feel that we're assuming the head should never change, and my worry is that other code might assume it (e.g. they might cache the value and assume it will never change).
While this is true for now, I guess in the future we might want to support something like `PrependProcessor(processor)` or `InsertProcessor(index, processor)`.

Reply:
@CodeBlanch do you know if the common practice in C# is to mark something `readonly` due to the semantics (e.g. we know that it will never change) or due to the actual usage (e.g. the fact that we didn't change it after initialization)?

To my knowledge in C++ it is focusing on semantics rather than the actual usage (e.g. const/mutable/volatile).

Reply:
@reyang I would say in C# it is based on actual usage. There is an [IDE analyzer](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0044) built into VS that suggests/warns about this case if it can tell you aren't modifying a private field. So you would have to go out of your way/fight the tooling (suppression, disable rule) to stick with the semantic approach.

Reply:
> I want to get more perspectives regarding this one.
> Making it `readonly` makes me feel that we're assuming the head should never change, and my worry is that other code might assume it (e.g. they might cache the value and assume it will never change).
> While this is true for now, I guess in the future we might want to support something like `PrependProcessor(processor)` or `InsertProcessor(index, processor)`.

As of now, the processors get added in the order of the `AddProcessor` calls. Would we need to offer something like `PrependProcessor` or `InsertProcessor` if we can just ask the user to order their `AddProcessor` calls accordingly?

Reply:
> As of now, the processors get added in the order of the `AddProcessor` calls. Would we need to offer something like `PrependProcessor` or `InsertProcessor` if we can just ask the user to order their `AddProcessor` calls accordingly?

I think in the world where a single user/dev/team controls everything, this should work, and is not an issue from the beginning.

In the real world, there might be separation of roles/responsibilities. For example you have an infrastructure team which developed several security or privacy related processors, and that team wants to "prepend" the security/privacy processor before other processors introduced by feature teams. This can be solved by having the infrastructure team establishing some rules (e.g. all the teams should call AddInfrastructureProcessors before adding their stuff), but practically it might be challenging.